### PR TITLE
feat(membership): join approval has membership decision + remove support for rejoins

### PR DESCRIPTION
**Pull request by davidrusu**
This PR changes our Join approval flow to include the membership decision that approved the node to join the section.

Also we remove support for rejoins since that turns out to be unsound

### CODE REVIEW

File: sn_interface/src/network_knowledge/section_authority_provider.rs [hunk 1]
```diff
@@ -339,6 +340,42 @@ pub mod test_utils {
         })
     }
 
+    pub fn section_decision<P: Proposition>(
```
Comment (davidrusu) [inline sn_interface/src/network_knowledge/section_authority_provider.rs:343]: Test helper for creating any `sn_consensus::Decision<T>`

File: sn_node/src/node/messaging/agreement.rs [hunk 2]
```diff
@@ -72,107 +41,15 @@ impl Node {
         }
     }
 
-    pub(crate) async fn handle_online_agreement(
```
Comment (davidrusu) [inline sn_node/src/node/messaging/agreement.rs:75]: This handler has been renamed to `handle_join_decision()` and moved to the `messaging/membership.rs` module.

File: sn_node/src/node/messaging/membership.rs [hunk 3]
```diff
@@ -184,4 +190,141 @@ impl Node {
...
+
+        // Do not disable node joins in first section.
+        let our_prefix = self.network_knowledge.prefix();
+        if !our_prefix.is_empty() {
+            // ..otherwise, switch off joins_allowed on a node joining.
+            // TODO: fix racing issues here? https://github.com/maidsafe/safe_network/issues/890
+            self.joins_allowed = false;
+        }
+
+        if let Some((_, sig)) = joining_nodes.iter().max_by_key(|(_, sig)| sig) {
```
Comment (davidrusu) [inline sn_node/src/node/messaging/membership.rs:224]: Now only one churn id is created per generation, previously a churn id was created for each node that joined in that generation.

We use the largest signature of the nodes that joined this generation as the ChurnId

**Review by joshuef — CHANGES_REQUESTED** [review]
Changes look good. Though there's some failing node tests there

File: sn_node/src/node/cfg/config_handler.rs [hunk 4]
```diff
@@ -135,12 +135,6 @@ impl Config {
     /// Returns a new `Config` instance.  Tries to read from the default node config file location,
     /// and overrides values with any equivalent cmd line args.
     pub async fn new() -> Result<Self, Error> {
-        // FIXME: Re-enable when we have rejoins working
-        // let mut config = match Self::read_from_file() {
-        //     Ok(Some(config)) => config,
-        //     Ok(None) | Err(_) => Default::default(),
-        // };
-
```
Comment (davidrusu) [inline sn_node/src/node/cfg/config_handler.rs:143]: We've dropped support for rejoins in this PR

File: sn_node/src/node/flow_ctrl/tests/mod.rs [hunk 5]
```diff
@@ -610,46 +600,41 @@ async fn handle_agreement_on_online_of_rejoined_node(phase: NetworkPhase, age: u
...
-#[tokio::test]
-async fn handle_agreement_on_online_of_rejoined_node_with_low_age_in_startup() -> Result<()> {
-    handle_agreement_on_online_of_rejoined_node(NetworkPhase::Startup, 8).await
-}
-
-#[tokio::test]
-async fn handle_agreement_on_online_of_rejoined_node_with_low_age_after_startup() -> Result<()> {
-    handle_agreement_on_online_of_rejoined_node(NetworkPhase::Regular, 8).await
-}
-
```
Comment (davidrusu) [inline sn_node/src/node/flow_ctrl/tests/mod.rs:652]: We no longer support rejoins as of this PR so the various scenarios have been reduced down to the one "ensure we don't let nodes rejoin" scenario above

File: sn_node/src/node/flow_ctrl/tests/mod.rs [hunk 6]
```diff
@@ -610,46 +600,41 @@ async fn handle_agreement_on_online_of_rejoined_node(phase: NetworkPhase, age: u
...
+                .propose_membership_change(node_state.clone())
+                .unwrap();
 
-            assert!(status.node_approval_sent);
-            assert_matches!(status.relocate_details, Some(details) => {
-                assert_eq!(details.dst, peer.name());
-                assert_eq!(details.age, (age / 2).max(MIN_ADULT_AGE));
-            });
-            Result::<()>::Ok(())
+            // A rejoining node with always be rejected
```
Comment (oetyng) [inline sn_node/src/node/flow_ctrl/tests/mod.rs:623]: with => will

**Review by oetyng — APPROVED** [review]

**Pull request merged**